power_grid_model_c/power_grid_model/include/power_grid_model/main_core/output.hpp

```diff
@@ -0,0 +1,287 @@
+requires model_component_state<MainModelState, ComponentContainer, Component>
+constexpr auto comp_base_sequence_cbegin(MainModelState<ComponentContainer> const& state) {
+    return state.comp_coup->source.cbegin();
+}
+
+template <std::derived_from<GenericLoadGen> Component, class ComponentContainer>
+requires model_component_state<MainModelState, ComponentContainer, Component>
+constexpr auto comp_base_sequence_cbegin(MainModelState<ComponentContainer> const& state) {
+    return state.comp_coup->load_gen.cbegin() + comp_sequence_offset<GenericLoadGen, Component>(state);
+}
+
+template <std::same_as<Shunt> Component, class ComponentContainer>
+requires model_component_state<MainModelState, ComponentContainer, Component>
+constexpr auto comp_base_sequence_cbegin(MainModelState<ComponentContainer> const& state) {
+    return state.comp_coup->shunt.cbegin();
+}
+
+template <std::derived_from<GenericVoltageSensor> Component, class ComponentContainer>
+requires model_component_state<MainModelState, ComponentContainer, Component>
+constexpr auto comp_base_sequence_cbegin(MainModelState<ComponentContainer> const& state) {
+    return state.comp_topo->voltage_sensor_node_idx.cbegin() +
+           comp_sequence_offset<GenericVoltageSensor, Component>(state);
+}
+
+template <std::derived_from<GenericPowerSensor> Component, class ComponentContainer>
+requires model_component_state<MainModelState, ComponentContainer, Component>
+constexpr auto comp_base_sequence_cbegin(MainModelState<ComponentContainer> const& state) {
+    return state.comp_topo->power_sensor_object_idx.cbegin() +
+           comp_sequence_offset<GenericPowerSensor, Component>(state);
+}
+
+template <std::same_as<Fault> Component, class ComponentContainer>
+requires model_component_state<MainModelState, ComponentContainer, Component>
+constexpr auto comp_base_sequence_cbegin(MainModelState<ComponentContainer> const& state) {
+    return state.comp_coup->fault.cbegin();
+}
+
+template <typename Component, typename IndexType, class ComponentContainer, std::forward_iterator ResIt,
+          typename ResFunc>
+requires model_component_state<MainModelState, ComponentContainer, Component> &&
+    std::invocable<std::remove_cvref_t<ResFunc>, Component const&, IndexType>&&
+        std::convertible_to<std::invoke_result_t<ResFunc, Component const&, IndexType>, std::iter_value_t<ResIt>>&&
+            std::convertible_to<IndexType, decltype(*comp_base_sequence_cbegin<Component>(
+                                               MainModelState<ComponentContainer>{}))> constexpr ResIt
+            produce_output(MainModelState<ComponentContainer> const& state, ResIt res_it, ResFunc&& func) {
+    return std::transform(state.components.template citer<Component>().begin(),
+                          state.components.template citer<Component>().end(),
+                          comp_base_sequence_cbegin<Component>(state), res_it, func);
```

Comment:
doing it this way ensures the logic between actual conversion (declared later), writing output (declared here) and declaring input (declared earlier in `comp_base_sequence_cbegin`) is split up.

i know it costs quite a lot more lines than initially, but compile time is actually faster than it was before because it can pre-compile all functions separately